app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,166 @@
+# frozen_string_literal: true
+
+class ComposerController < ApplicationController
+  requires_login
+
+  def mentions
+    names = params.require(:names)
+    raise Discourse::InvalidParameters.new(:names) if !names.kind_of?(Array) || names.size > 20
+
+    if params[:topic_id].present?
+      topic = Topic.find_by(id: params[:topic_id])
+      guardian.ensure_can_see!(topic)
+    end
+
+    # allowed_names is necessary just for new private messages.
+    allowed_names = if params[:allowed_names].present?
+      params[:allowed_names].split(',') << current_user.username
+    else
+      []
+    end
+
+    user_reasons = {}
+    group_reasons = {}
+    here_count = nil
+
+    users = User
+      .not_staged
+      .where(username_lower: names.map(&:downcase))
+      .index_by(&:username_lower)
+
+    groups = Group
+      .visible_groups(current_user)
+      .where(name: names)
+      .index_by(&:name)
+
+    mentionable_group_ids = Group
+      .mentionable(current_user, include_public: false)
+      .where(name: names)
+      .pluck(:id)
+      .to_set
+
+    members_visible_group_ids = Group
+      .members_visible_groups(current_user)
+      .where(name: names)
+      .pluck(:id)
+      .to_set
+
+    topic_muted_by = if topic.present?
+      TopicUser
+        .where(topic: topic)
+        .where(user_id: users.values.map(&:id))
+        .where(notification_level: TopicUser.notification_levels[:muted])
+        .pluck(:user_id)
+        .to_set
+    else
+      Set.new
+    end
+
+    topic_allowed_user_ids = if allowed_names.present?
+      User
+        .where(username_lower: allowed_names.map(&:downcase))
+        .pluck(:id)
+        .to_set
+    elsif topic&.private_message?
+      TopicAllowedUser
+        .where(topic: topic)
+        .pluck(:user_id)
+        .to_set
+    end
+
+    topic_allowed_group_ids = if allowed_names.present?
+      Group
+        .messageable(current_user)
+        .where(name: allowed_names)
+        .pluck(:id)
+        .to_set
+    elsif topic&.private_message?
+      TopicAllowedGroup
+        .where(topic: topic)
+        .pluck(:group_id)
+        .to_set
+    end
```

Comment:
```suggestion
    if allowed_names.present?
      topic_allowed_user_ids = User.where(username_lower: allowed_names.map(&:downcase)).distinct(:id).pluck(:id)
      topic_allowed_group_ids = Group.messageable(current_user).where(name: allowed_names).distinct(:id).pluck(:id)
    elsif topic&.private_message?
      topic_allowed_user_ids = TopicAllowedUser.where(topic: topic).distinct(:user_id).pluck(:user_id)
      topic_allowed_group_ids = TopicAllowedGroup.where(topic: topic).distinct(:group_id).pluck(:group_id)
    end
```

Small suggestion to avoid having to duplicate the conditions. Also, we can just pluck distinct instead of having to call `.to_set` on the results.

Reply:
I will move them to different fields, so I will need to keep the duplicate conditions.

The `to_set` is not to avoid duplication, but to have fast lookups.